### PR TITLE
Make RemainingItemCount a pointer

### LINKED ISSUE #78318: RemainingItemCount is serialized in proto messages even when unset
**What happened**:
As part of [adding skew testing for serialized API responses](https://github.com/kubernetes/kubernetes/pull/78309), we discovered a diff in serialized proto responses when roundtripping data from 1.14:

```
    --- FAIL: TestRoundTripExternalTypes/.v1.Status (0.02s)
        --- FAIL: TestRoundTripExternalTypes/.v1.Status/fixtures-v1.14.0 (0.02s)
            fixtures.go:195: proto data differed from fixture
            fixtures.go:205: fixture vs reserialized:
                  strings.Join({
                  	... // 7 identical lines
                  	`    2: "8551505577692297499"`,
                  	`    3: ""`,
                + 	"    4: 0",
                  	"  }",
                  	"  2 {",
                  	... // 21 identical lines
                  }, "\n")
```

This is due to the RemainingItemCount field in ListMeta, which is always serialized in proto messages, because it is not a pointer.

**What you expected to happen**:
Proto messages do not include the optional field when unset.

**How to reproduce it (as minimally and precisely as possible)**:
* Check out https://github.com/kubernetes/kubernetes/pull/78309
* Run `go test ./vendor/k8s.io/api -run TestRoundTripExternalTypes//1.14` (tests round-tripping serialized API content for all types we shipped in 1.14)

This needs to be resolved before 1.15 ships.

/milestone v1.15
/kind bug
/sig api-machinery
/priority important-soon
/assign @caesarxuchao

### PR DESCRIPTION
**Pull request by caesarxuchao**
/kind bug
/sig api-machinery
/assign @liggitt 

fixes https://github.com/kubernetes/kubernetes/issues/78318

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/meta.go [hunk 1]
```diff
@@ -115,8 +115,20 @@ func (meta *ListMeta) GetSelfLink() string               { return meta.SelfLink
...
+func (meta *ListMeta) GetRemainingItemCount() int64 {
+	if meta.RemainingItemCount != nil {
+		return *meta.RemainingItemCount
+	}
+	return 0
+}
+
+func (meta *ListMeta) SetRemainingItemCount(c int64) {
+	if meta.RemainingItemCount == nil {
+		meta.RemainingItemCount = new(int64)
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/meta.go]: this implementation means that all calls to SetRemainingItemCount will result in a serialized output. If we need to distinguish between set and unset, we should have this:
```
GetRemainingItemCount() *int64
SetRemainingItemCount(c *int64)
```

and calling `SetRemainingItemCount(nil)` should remove the field from unstructured implementations

File: staging/src/k8s.io/apimachinery/pkg/test/api_meta_meta_test.go [hunk 2]
```diff
@@ -202,22 +201,34 @@ type InternalTypeMeta struct {
...
+func (m *InternalTypeMeta) GetRemainingItemCount() int64 {
+	if m.RemainingItemCount != nil {
+		return *m.RemainingItemCount
+	}
+	return 0
+}
+
+func (m *InternalTypeMeta) SetRemainingItemCount(c int64) {
+	if m.RemainingItemCount == nil {
+		m.RemainingItemCount = new(int64)
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/test/api_meta_meta_test.go]: same comment here, SetRemainingItemCount(nil) should clear the field

File: staging/src/k8s.io/apiserver/pkg/registry/rest/table.go [hunk 3]
```diff
@@ -67,7 +67,7 @@ func (c defaultTableConvertor) ConvertToTable(ctx context.Context, object runtim
 		table.ResourceVersion = m.GetResourceVersion()
 		table.SelfLink = m.GetSelfLink()
 		table.Continue = m.GetContinue()
-		table.RemainingItemCount = m.GetRemainingItemCount()
+		table.SetRemainingItemCount(m.GetRemainingItemCount())
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/registry/rest/table.go]: this is why we need to make sure nil can be propagated from GetRemainingItemCount to SetRemainingItemCount without forcing a 0 to be serialized

**Comment by caesarxuchao** [comment]
@liggitt sorry I won't have time to fix this PR until May 29. I have enabled edits for maintainers. Also feel free to supersede this PR if necessary.

**Comment by liggitt** [comment]
updated to propagate use of the pointers to the listmeta interface

**Pull request merged**